Native indexes recovery cleanup.

MishaDemianenko opened the pull request:
Do not try to recover native label scan store on CC.
Allow native scan store to shutdown even if init failed.
Cleanup in RecoverRequiredChecker.

MishaDemianenko committed:
Native indexes recovery cleanup.

Do not try to recover native label scan store on CC.
Allow native scan store to shutdown even if init failed.
Cleanup in RecoverRequiredChecker.

burqen commented on community/consistency-check/src/main/java/org/neo4j/consistency/ConsistencyCheckService.java:
```diff
@@ -243,7 +242,7 @@ public Result runFullConsistencyCheck( final File storeDir, Config config, Progr
             LabelScanStore labelScanStore =
                     new NativeLabelScanStore( pageCache, storeDir, FullStoreChangeStream.EMPTY, true, monitors,
-                            RecoveryCleanupWorkCollector.IMMEDIATE );
+                            RECOVERY_PREVENTING_COLLECTOR );
```
First of all, it is clear that we should never execute the cleanup job because consistency check should be a read only thing. This change I like!
However, I don't think it is the right thing to block consistency checker from doing it's job since we know that we are not doing a "real" recovery of the store. At the same time, I don't think silently ignoring is a good thing either. Perhaps we could report a dirty tree as any other inconsistency? What do you think?
If this turns out to be a big change, it can wait for some future PR.

burqen reviewed (changes requested):
Nice cleanups!
It's not super clear to be what would be the correct behaviour. See comment and let's discuss.

burqen commented:
@MishaDemianenko Did we get anywhere with this?

MishaDemianenko commented:
@burqen sorry didn't have time to come back into this, will try to do things as we agreed as soon as i will switch back into it

MishaDemianenko committed:
Allow consistency checker to run on top of not clean native indexes

Ignore recovery for native indexes during consistency check instead
of throwing exception that indexes are dirty.
Also do not rebuild label index if it was open in read only mode.

MishaDemianenko committed:
Allow consistency checker to warn about indexes that are dirty
and can return wrong results

Dirty native index is an index that will require recovery on next non
read only startup and potentially any data that is in the index
on a moment when index is dirty are not fully trustable.

This commit allow CC to notify about those indexes during CC check and
report those.

MishaDemianenko commented:
@burqen introduced CC checks for the dirty label and native indexes. Not sure I'm in the fan zone for that kind of check, but let's see wdyt.

burqen commented on community/consistency-check/src/test/java/org/neo4j/consistency/checking/AllNodesInStoreExistInLabelIndexTest.java:
```diff
@@ -83,6 +86,31 @@ public void mustReportSuccessfulForConsistentLabelScanStore() throws Exception
     }
 
+    @Test
+    public void reportNotCleanLabelIndex() throws IOException, ConsistencyCheckIncompleteException
```
This test a label scan store that is actually missing some label and thus "really" is corrupt. We also want to test the situation where label index actually contain the correct information but is dirty.

Something like this:
```
    @Test
    public void reportNotCleanLabelIndexWithCorrectData() throws IOException, ConsistencyCheckIncompleteException
    {
        File storeDir = db.getStoreDir();
        someData();
        db.resolveDependency( CheckPointer.class ).forceCheckPoint( new SimpleTriggerInfo( "forcedCheckpoint" ) );
        File labelIndexFileCopy = new File( storeDir, "label_index_copy" ); // <- COPIED FILE HAS ALL THE STUFFZ

        copyFile( new File( storeDir, NativeLabelScanStore.FILE_NAME ), labelIndexFileCopy );

        db.shutdownAndKeepStore();

        copyFile( labelIndexFileCopy, new File( storeDir, NativeLabelScanStore.FILE_NAME ) );

        ConsistencyCheckService.Result result = fullConsistencyCheck();
        assertTrue( "Expected consistency check to fail", result.isSuccessful() ); // <- SUCCESSFUL!
        assertThat( readReport( result ),
                hasItem( containsString("WARN : Label index was not properly shutdown and rebuild is required.") ) );
    }
```

burqen commented on community/consistency-check/src/test/java/org/neo4j/consistency/checking/IndexConsistencyIT.java:
```diff
@@ -0,0 +1,224 @@
+    private static final int NODE_COUNT_BASELINE = 10;
+
+    @Test
+    public void reportNotCleanNativeIndex() throws IOException, ConsistencyCheckIncompleteException
```
Same thing goes for this test

burqen commented on community/index/src/main/java/org/neo4j/index/internal/gbptree/GBPTree.java:
```diff
@@ -1297,4 +1301,9 @@ private void closeCursor()
         }
     }
+
+    public boolean isDirty()
```
Misleading name since state could have changed since startup, `wasDirtyOnStartup` or something to that effect. Also, this functionality already exist through the `Monitor`. If we want to rely on that we can edit `private GBPTree.Monitor treeMonitor()` in `NativeLabelScanStore` and `NativeSchemaNumberIndex`.

MishaDemianenko replied:
no i will not gonna use monitor for that, its not a monitor business to return you some kind of object state, that just wrong

burqen replied:
Alright, just the name change then 😄

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/labelscan/NativeLabelScanStore.java:
```diff
@@ -467,12 +470,20 @@ public void stop() throws IOException
-        index.close();
+        if ( index != null )
+        {
+            index.close();
```
Maybe take the opportunity to set `index = null` here?

burqen commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/labelscan/NativeLabelScanStoreRebuildTest.java:
```diff
@@ -108,6 +95,46 @@ public void mustBeDirtyIfFailedDuringRebuild() throws Exception
+    }
+
+    @Test
+    public void labelScanStoreIdDirtyWhenIndexIsNotClean() throws IOException
```
labelScanStore**Is**DirtyWhenIndexIsNotClean

burqen reviewed (changes requested):
I think this is a quite nice compromise between "kill checkpoint on dirty state" and "silently allow checkpoint even if state is dirty".

Found some improvements here and there.

MishaDemianenko committed:
Review comments:
additional tests for dirty index without missing data;
renaming;
mistype fixes,
nullify index field.

MishaDemianenko commented:
@burqen here we go. Comments should be resolved and also includes `the renaming` :)
